plugin-server/src/worker/ingestion/event-pipeline/1-populateTeamDataStep.ts

```diff
@@ -0,0 +1,31 @@
+import { PluginEvent } from '@posthog/plugin-scaffold'
+
+import { PipelineEvent } from '../../../types'
+import { EventPipelineRunner, StepResult } from './runner'
+
+// TRICKY: team_id is optional on PipelineEvent but not for PluginEvent
+// Only do the type asserton to PluginEvent having verified team_id exists
+export async function populateTeamDataStep(runner: EventPipelineRunner, event: PipelineEvent): Promise<StepResult> {
+    if (event.team_id) {
+        return runner.nextStep('emitToBufferStep', event as PluginEvent)
+    }
+
+    if (!event.token) {
+        runner.hub.statsd?.increment('dropped_event_with_no_team', { token_set: 'false' })
+        return null
+    }
+
+    const team = await runner.hub.teamManager.fetchTeam(null, event.token)
+    if (!team) {
+        runner.hub.statsd?.increment('dropped_event_with_no_team', { token_set: 'true' })
+        return null
+    }
+
+    event = {
+        ...event,
+        team_id: team.id,
+        ip: team.anonymize_ips ? null : event.ip,
```

Comment:
It's not obvious why we're doing this here as we exit out of this step early if event.team_id is set.

Perhaps only plugin-server should be unsetting ips? Alternatively, we should strongly document this implicit contract (if event.team_id is set, capture endpoint is responsible for this business logic).

Reply:
it does make sense to me - we need to do this before any plugins are ran. Longer term the plan is all events go through this, so it seems fine to do it here. Though from a structural point I wouldn't expect `populateTeamDataStep` to do anything with ips, so maybe the file name should be something else indicating pre-processing of events.

Please update the docs accordingly, e.g. https://posthog.com/docs/runbook/services/plugin-server#worker-thread is wrong (could link to the how posthog works one) ; https://posthog.com/docs/how-posthog-works/ingestion-pipeline#plugin-server needs updating